### ISSUE TITLE
Fail to mock in 3.2.0, but succeed in 3.1.0

### ISSUE THREAD

**Issue by bingosummer** [rspec/rspec-mocks#1161]
When I upgrade rspec from `3.1.0` to `3.2.0`, I hit the following issue.

I mocked a class using `let(:client) { instance_double(Azure::Storage::Client) }`, and mocked one method `allow(client).to receive(:storage_table_host=)`. When I run `bundle exec rspec`, it fails with an error.
```
     Failure/Error: allow(client).to receive(:storage_table_host=)
     Azure::Storage::InvalidOptionsError:
       options provided are not valid set: {}
```

`Azure::Storage::InvalidOptionsError` is defined in `Azure::Storage::Client`. From my understanding, if I mocked the class successfully, I should not get this kind of error.

FYI; The bundle env:
```
Bundler   1.15.0
Rubygems  2.4.5.1
Ruby      2.2.4p230 (2015-12-16 revision 53155) [x86_64-linux]
Git       2.1.4
Platform  x86_64-linux

...

  gem 'rspec', '~>3.2.0'
  gem 'rake', '~>10.3.2'
```

Do you have any suggestions? Thanks.

**Comment by myronmarston** [comment]
What gem does `Azure::Storage` come from and what version?

**Comment by bingosummer** [comment]
`azure-storage (= 0.12.1.preview)`

**Comment by myronmarston** [comment]
There's a bug in `azure-storage`.

Normally, code like `[klass, 'SubModule'].join('::')` works:

``` irb
irb(main):004:0> [Array, 'SubClass'].join('::')
=> "Array::SubClass"
```

But `Azure::Storage` is busted:

``` irb
irb(main):003:0> [Azure::Storage, 'Client'].join('::')
Azure::Storage::InvalidOptionsError: options provided are not valid set: {}
	from /Users/myron/code/rspec-scratch/rspec-mocks-1161/bundle/ruby/2.4.0/gems/azure-storage-0.12.1.preview/lib/azure/storage/client_options.rb:253:in `filter'
	from /Users/myron/code/rspec-scratch/rspec-mocks-1161/bundle/ruby/2.4.0/gems/azure-storage-0.12.1.preview/lib/azure/storage/client_options.rb:86:in `reset!'
	from /Users/myron/code/rspec-scratch/rspec-mocks-1161/bundle/ruby/2.4.0/gems/azure-storage-0.12.1.preview/lib/azure/storage/client.rb:89:in `initialize'
	from /Users/myron/code/rspec-scratch/rspec-mocks-1161/bundle/ruby/2.4.0/gems/azure-storage-0.12.1.preview/lib/azure/storage.rb:41:in `new'
	from /Users/myron/code/rspec-scratch/rspec-mocks-1161/bundle/ruby/2.4.0/gems/azure-storage-0.12.1.preview/lib/azure/storage.rb:41:in `client'
	from /Users/myron/code/rspec-scratch/rspec-mocks-1161/bundle/ruby/2.4.0/gems/azure-storage-0.12.1.preview/lib/azure/storage.rb:49:in `method_missing'
```

I think we can work around this, but `Azure::Storage` is a broken class and should be fixed. Please report to the maintainers.

**Comment by myronmarston** [comment]
I've got a fix in rspec/rspec-support#1161 if you want to give that a shot.

**Comment by bingosummer** [comment]
Thank you, @myronmarston. The fix works for me.

**Comment by katmsft** [comment]
Fixed in 0.14.0-preview of azure-storage-ruby.
